{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "62cd4ff2-c81b-42b2-814d-c4fd2016d9af",
   "metadata": {},
   "outputs": [],
   "source": [
    "from HWG.build import *\n",
    "import pandas as pd\n",
    "import os\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "f55471ab-39e4-488c-ad9b-e65fe652e146",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "****************\n",
      "Downloading Data\n",
      "****************\n",
      "    HuRI Downloaded\n",
      "    STRING Downloaded\n",
      "    HumanTF Downloaded\n",
      "    HGNC Downloaded\n",
      "********************\n",
      "Building Index Table\n",
      "********************\n",
      "    Missing TF: ENSG00000277462\n",
      "    Missing TF: ENSG00000214534\n",
      "    Missing TF: ENSG00000223614\n",
      "    Missing TF: ENSG00000228623\n",
      "    Missing TF: ENSG00000213793\n",
      "    Missing TF: DUX1_HUMAN\n",
      "    Missing TF: DUX3_HUMAN\n",
      "    Missing TF: ENSG00000260027\n",
      "    Missing TF: ENSG00000255192\n",
      "    Missing TF: ENSG00000230257\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jpic/HardwiredGenome-Python/HWG/build.py:306: DtypeWarning: Columns (32,34,38,40,50) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  HGNC_genes = pd.read_csv(HGNC_downloaded_data, sep='\\t')\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "********************\n",
      "Building Adj. Matrix\n",
      "********************\n",
      "HuRI Data Accessed\n",
      "Accessing STRING data\n",
      "    Reading data from file...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jpic/HardwiredGenome-Python/HWG/build.py:360: FutureWarning: The 'delim_whitespace' keyword in pd.read_csv is deprecated and will be removed in a future version. Use ``sep='\\s+'`` instead\n",
      "  STRING_PPI = pd.read_csv(STRING_downloaded_PPI, delim_whitespace=True)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    File read complete\n",
      "    93.45% of the data was removed because it is below the confidence threshold.\n",
      "    2.06% of the data was removed because it wasn't mapped to a stable gene ID.\n",
      "STRING Data Access Complete: 6.42% of the data was used.\n",
      "STRING Data Accessed\n",
      "********************\n",
      "   save to a file   \n",
      "********************\n",
      "path=/home/jpic/HardwiredGenome-Python/Data/Processed/HWG_600.pkl\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'thresh': 600,\n",
       " 'A': <41323x41323 sparse matrix of type '<class 'numpy.int64'>'\n",
       " \twith 884971 stored elements in Compressed Sparse Row format>,\n",
       " 'geneIndexTable':              Stable ID     symbol  Transcription Factor  Gene start (bp)  \\\n",
       " 0      ENSG00000000003     TSPAN6                 False      100627108.0   \n",
       " 1      ENSG00000000005       TNMD                 False      100584936.0   \n",
       " 2      ENSG00000000419       DPM1                 False       50934867.0   \n",
       " 3      ENSG00000000457      SCYL3                 False              NaN   \n",
       " 4      ENSG00000000460      FIRRM                 False              NaN   \n",
       " ...                ...        ...                   ...              ...   \n",
       " 41318  ENSG00000181404     WASHC1                 False              NaN   \n",
       " 41319  ENSG00000276581  SPATA31A5                 False              NaN   \n",
       " 41320  ENSG00000278848   TP53TG3F                 False              NaN   \n",
       " 41321  ENSG00000279782    PPIAL4F                 False              NaN   \n",
       " 41322  ENSG00000280267   PRAMEF26                 False              NaN   \n",
       " \n",
       "        Gene end (bp) Chromosome/scaffold name Protein stable ID  \n",
       " 0        100639991.0                        X   ENSP00000362111  \n",
       " 1        100599885.0                        X   ENSP00000362122  \n",
       " 2         50959140.0                       20   ENSP00000507119  \n",
       " 3                NaN                      NaN               NaN  \n",
       " 4                NaN                      NaN               NaN  \n",
       " ...              ...                      ...               ...  \n",
       " 41318            NaN                      NaN               NaN  \n",
       " 41319            NaN                      NaN               NaN  \n",
       " 41320            NaN                      NaN               NaN  \n",
       " 41321            NaN                      NaN               NaN  \n",
       " 41322            NaN                      NaN               NaN  \n",
       " \n",
       " [41323 rows x 7 columns]}"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "buildHWG()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "8149a67a-7128-417b-a446-c813866600bc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HWG for threshold 0 does not exist.\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Do you want to build HWG with this threshold? (yes/no):  yes\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "****************\n",
      "Downloading Data\n",
      "****************\n",
      "    HuRI Downloaded\n",
      "    STRING Downloaded\n",
      "    HumanTF Downloaded\n",
      "    HGNC Downloaded\n",
      "********************\n",
      "Building Index Table\n",
      "********************\n",
      "    Missing TF: ENSG00000277462\n",
      "    Missing TF: ENSG00000214534\n",
      "    Missing TF: ENSG00000223614\n",
      "    Missing TF: ENSG00000228623\n",
      "    Missing TF: ENSG00000213793\n",
      "    Missing TF: DUX1_HUMAN\n",
      "    Missing TF: DUX3_HUMAN\n",
      "    Missing TF: ENSG00000260027\n",
      "    Missing TF: ENSG00000255192\n",
      "    Missing TF: ENSG00000230257\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jpic/HardwiredGenome-Python/HWG/build.py:306: DtypeWarning: Columns (32,34,38,40,50) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  HGNC_genes = pd.read_csv(HGNC_downloaded_data, sep='\\t')\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "********************\n",
      "Building Adj. Matrix\n",
      "********************\n",
      "HuRI Data Accessed\n",
      "Accessing STRING data\n",
      "    Reading data from file...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jpic/HardwiredGenome-Python/HWG/build.py:360: FutureWarning: The 'delim_whitespace' keyword in pd.read_csv is deprecated and will be removed in a future version. Use ``sep='\\s+'`` instead\n",
      "  STRING_PPI = pd.read_csv(STRING_downloaded_PPI, delim_whitespace=True)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    File read complete\n",
      "    0.00% of the data was removed because it is below the confidence threshold.\n",
      "    2.08% of the data was removed because it wasn't mapped to a stable gene ID.\n",
      "STRING Data Access Complete: 97.92% of the data was used.\n",
      "STRING Data Accessed\n",
      "********************\n",
      "   save to a file   \n",
      "********************\n",
      "path=/home/jpic/HardwiredGenome-Python/Data/Processed/HWG_0.pkl\n"
     ]
    }
   ],
   "source": [
    "HWG = loadHWG(thresh=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "d6acfa7f-1efd-4614-8572-34689d8b5f3e",
   "metadata": {},
   "outputs": [],
   "source": [
    "path = '/nfs/turbo/umms-indikar/shared/projects/public_data/HWG/HWG0.pkl'\n",
    "with open(path, 'wb') as f:\n",
    "    pickle.dump(HWG, f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "7bc58fc9-dd76-40fd-ae03-dd2d97201511",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['thresh', 'A', 'geneIndexTable'])"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "HWG.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "d66d7241-6cae-4e18-9f25-25484c497829",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "HWG['thresh']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "cf23b883-4caa-401e-b90b-96f0daa073cc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<41323x41323 sparse matrix of type '<class 'numpy.int64'>'\n",
       "\twith 11750266 stored elements in Compressed Sparse Row format>"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "HWG['A']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "64056774-1e0c-4f8c-96c9-8445eccd96cf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Stable ID</th>\n",
       "      <th>symbol</th>\n",
       "      <th>Transcription Factor</th>\n",
       "      <th>Gene start (bp)</th>\n",
       "      <th>Gene end (bp)</th>\n",
       "      <th>Chromosome/scaffold name</th>\n",
       "      <th>Protein stable ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>False</td>\n",
       "      <td>100627108.0</td>\n",
       "      <td>100639991.0</td>\n",
       "      <td>X</td>\n",
       "      <td>ENSP00000362111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000000005</td>\n",
       "      <td>TNMD</td>\n",
       "      <td>False</td>\n",
       "      <td>100584936.0</td>\n",
       "      <td>100599885.0</td>\n",
       "      <td>X</td>\n",
       "      <td>ENSP00000362122</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000000419</td>\n",
       "      <td>DPM1</td>\n",
       "      <td>False</td>\n",
       "      <td>50934867.0</td>\n",
       "      <td>50959140.0</td>\n",
       "      <td>20</td>\n",
       "      <td>ENSP00000507119</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000000457</td>\n",
       "      <td>SCYL3</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000000460</td>\n",
       "      <td>FIRRM</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41318</th>\n",
       "      <td>ENSG00000181404</td>\n",
       "      <td>WASHC1</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41319</th>\n",
       "      <td>ENSG00000276581</td>\n",
       "      <td>SPATA31A5</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41320</th>\n",
       "      <td>ENSG00000278848</td>\n",
       "      <td>TP53TG3F</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41321</th>\n",
       "      <td>ENSG00000279782</td>\n",
       "      <td>PPIAL4F</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41322</th>\n",
       "      <td>ENSG00000280267</td>\n",
       "      <td>PRAMEF26</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>41323 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Stable ID     symbol  Transcription Factor  Gene start (bp)  \\\n",
       "0      ENSG00000000003     TSPAN6                 False      100627108.0   \n",
       "1      ENSG00000000005       TNMD                 False      100584936.0   \n",
       "2      ENSG00000000419       DPM1                 False       50934867.0   \n",
       "3      ENSG00000000457      SCYL3                 False              NaN   \n",
       "4      ENSG00000000460      FIRRM                 False              NaN   \n",
       "...                ...        ...                   ...              ...   \n",
       "41318  ENSG00000181404     WASHC1                 False              NaN   \n",
       "41319  ENSG00000276581  SPATA31A5                 False              NaN   \n",
       "41320  ENSG00000278848   TP53TG3F                 False              NaN   \n",
       "41321  ENSG00000279782    PPIAL4F                 False              NaN   \n",
       "41322  ENSG00000280267   PRAMEF26                 False              NaN   \n",
       "\n",
       "       Gene end (bp) Chromosome/scaffold name Protein stable ID  \n",
       "0        100639991.0                        X   ENSP00000362111  \n",
       "1        100599885.0                        X   ENSP00000362122  \n",
       "2         50959140.0                       20   ENSP00000507119  \n",
       "3                NaN                      NaN               NaN  \n",
       "4                NaN                      NaN               NaN  \n",
       "...              ...                      ...               ...  \n",
       "41318            NaN                      NaN               NaN  \n",
       "41319            NaN                      NaN               NaN  \n",
       "41320            NaN                      NaN               NaN  \n",
       "41321            NaN                      NaN               NaN  \n",
       "41322            NaN                      NaN               NaN  \n",
       "\n",
       "[41323 rows x 7 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "HWG['geneIndexTable']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "a621a84d-a932-460f-88f5-decf5e578af0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/home/jpic/HardwiredGenome-Python/Data/Processed/HWG_600.pkl'"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "thresh = 600\n",
    "fileName = 'HWG_' + str(thresh) + '.pkl'\n",
    "path = load_data_path()\n",
    "path = os.path.join(path, 'Processed', fileName)\n",
    "path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "ed552a91-15c6-43f6-974c-c626690d8575",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    Missing TF: ENSG00000277462\n",
      "    Missing TF: ENSG00000214534\n",
      "    Missing TF: ENSG00000223614\n",
      "    Missing TF: ENSG00000228623\n",
      "    Missing TF: ENSG00000213793\n",
      "    Missing TF: DUX1_HUMAN\n",
      "    Missing TF: DUX3_HUMAN\n",
      "    Missing TF: ENSG00000260027\n",
      "    Missing TF: ENSG00000255192\n",
      "    Missing TF: ENSG00000230257\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jpic/HardwiredGenome-Python/HWG/build.py:240: DtypeWarning: Columns (32,34,38,40,50) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  HGNC_genes = pd.read_csv(HGNC_downloaded_data, sep='\\t')\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Stable ID</th>\n",
       "      <th>symbol</th>\n",
       "      <th>Transcription Factor</th>\n",
       "      <th>Gene start (bp)</th>\n",
       "      <th>Gene end (bp)</th>\n",
       "      <th>Chromosome/scaffold name</th>\n",
       "      <th>Protein stable ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>False</td>\n",
       "      <td>100627108.0</td>\n",
       "      <td>100639991.0</td>\n",
       "      <td>X</td>\n",
       "      <td>ENSP00000362111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000000005</td>\n",
       "      <td>TNMD</td>\n",
       "      <td>False</td>\n",
       "      <td>100584936.0</td>\n",
       "      <td>100599885.0</td>\n",
       "      <td>X</td>\n",
       "      <td>ENSP00000362122</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000000419</td>\n",
       "      <td>DPM1</td>\n",
       "      <td>False</td>\n",
       "      <td>50934867.0</td>\n",
       "      <td>50959140.0</td>\n",
       "      <td>20</td>\n",
       "      <td>ENSP00000507119</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000000457</td>\n",
       "      <td>SCYL3</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000000460</td>\n",
       "      <td>FIRRM</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41318</th>\n",
       "      <td>ENSG00000181404</td>\n",
       "      <td>WASHC1</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41319</th>\n",
       "      <td>ENSG00000276581</td>\n",
       "      <td>SPATA31A5</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41320</th>\n",
       "      <td>ENSG00000278848</td>\n",
       "      <td>TP53TG3F</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41321</th>\n",
       "      <td>ENSG00000279782</td>\n",
       "      <td>PPIAL4F</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41322</th>\n",
       "      <td>ENSG00000280267</td>\n",
       "      <td>PRAMEF26</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>41323 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Stable ID     symbol  Transcription Factor  Gene start (bp)  \\\n",
       "0      ENSG00000000003     TSPAN6                 False      100627108.0   \n",
       "1      ENSG00000000005       TNMD                 False      100584936.0   \n",
       "2      ENSG00000000419       DPM1                 False       50934867.0   \n",
       "3      ENSG00000000457      SCYL3                 False              NaN   \n",
       "4      ENSG00000000460      FIRRM                 False              NaN   \n",
       "...                ...        ...                   ...              ...   \n",
       "41318  ENSG00000181404     WASHC1                 False              NaN   \n",
       "41319  ENSG00000276581  SPATA31A5                 False              NaN   \n",
       "41320  ENSG00000278848   TP53TG3F                 False              NaN   \n",
       "41321  ENSG00000279782    PPIAL4F                 False              NaN   \n",
       "41322  ENSG00000280267   PRAMEF26                 False              NaN   \n",
       "\n",
       "       Gene end (bp) Chromosome/scaffold name Protein stable ID  \n",
       "0        100639991.0                        X   ENSP00000362111  \n",
       "1        100599885.0                        X   ENSP00000362122  \n",
       "2         50959140.0                       20   ENSP00000507119  \n",
       "3                NaN                      NaN               NaN  \n",
       "4                NaN                      NaN               NaN  \n",
       "...              ...                      ...               ...  \n",
       "41318            NaN                      NaN               NaN  \n",
       "41319            NaN                      NaN               NaN  \n",
       "41320            NaN                      NaN               NaN  \n",
       "41321            NaN                      NaN               NaN  \n",
       "41322            NaN                      NaN               NaN  \n",
       "\n",
       "[41323 rows x 7 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "indexTable = buildIndexTable()\n",
    "indexTable"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "bc6b76f6-9044-4b03-bd4f-0edc5a4f8a6d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HuRI Data Accessed\n",
      "Accessing STRING data\n",
      "    Reading data from file...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "KeyboardInterrupt\n",
      "\n"
     ]
    }
   ],
   "source": [
    "thresh = 600\n",
    "\n",
    "# Get adjacency lists\n",
    "A_list_HuRI = list_HuRI()\n",
    "print(\"HuRI Data Accessed\")\n",
    "\n",
    "A_list_STRING, _ = list_STRING(thresh)\n",
    "print(\"STRING Data Accessed\")\n",
    "\n",
    "# Combine the lists\n",
    "A_list = list_combine([A_list_HuRI, A_list_STRING])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "772071e8-2d9a-455e-955a-2cf0bcec4b6b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene 1</th>\n",
       "      <th>Gene 2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000000005</td>\n",
       "      <td>ENSG00000061656</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000000005</td>\n",
       "      <td>ENSG00000099968</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000000005</td>\n",
       "      <td>ENSG00000104765</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000000005</td>\n",
       "      <td>ENSG00000105383</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000000005</td>\n",
       "      <td>ENSG00000114455</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>830279</th>\n",
       "      <td>ENSG00000100129</td>\n",
       "      <td>ENSG00000114867</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>830280</th>\n",
       "      <td>ENSG00000100129</td>\n",
       "      <td>ENSG00000104408</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>830281</th>\n",
       "      <td>ENSG00000100129</td>\n",
       "      <td>ENSG00000116251</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>830282</th>\n",
       "      <td>ENSG00000100129</td>\n",
       "      <td>ENSG00000136628</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>830283</th>\n",
       "      <td>ENSG00000100129</td>\n",
       "      <td>ENSG00000140988</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>830284 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 Gene 1           Gene 2\n",
       "0       ENSG00000000005  ENSG00000061656\n",
       "1       ENSG00000000005  ENSG00000099968\n",
       "2       ENSG00000000005  ENSG00000104765\n",
       "3       ENSG00000000005  ENSG00000105383\n",
       "4       ENSG00000000005  ENSG00000114455\n",
       "...                 ...              ...\n",
       "830279  ENSG00000100129  ENSG00000114867\n",
       "830280  ENSG00000100129  ENSG00000104408\n",
       "830281  ENSG00000100129  ENSG00000116251\n",
       "830282  ENSG00000100129  ENSG00000136628\n",
       "830283  ENSG00000100129  ENSG00000140988\n",
       "\n",
       "[830284 rows x 2 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "A_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "c2e0326e-8ff4-4c57-ba1b-a9618ccc8053",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Build the matrix\n",
    "A_HWG = list2mat(A_list, indexTable)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "c3b64f4d-2a07-46bc-a3d3-7f4bf208c4e6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<41323x41323 sparse matrix of type '<class 'numpy.int64'>'\n",
       "\twith 884971 stored elements in Compressed Sparse Row format>"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "A_HWG"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "7e436b66-6b71-4536-b4d2-59e925c1ca3f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HuRI Data Accessed\n",
      "Accessing STRING data\n",
      "    Reading data from file...\n",
      "    File read complete\n",
      "    93.45% of the data was removed because it is below the confidence threshold.\n",
      "    2.06% of the data was removed because it wasn't mapped to a stable gene ID.\n",
      "STRING Data Access Complete: 6.42% of the data was used.\n",
      "STRING Data Accessed\n"
     ]
    }
   ],
   "source": [
    "HWG = buildAdjacencyMatrix(indexTable)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "096f086f-07bc-426e-b44e-795276c2ab44",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'thresh': 600,\n",
       " 'A': <41323x41323 sparse matrix of type '<class 'numpy.int64'>'\n",
       " \twith 884971 stored elements in Compressed Sparse Row format>,\n",
       " 'geneIndexTable':              Stable ID     symbol  Transcription Factor  Gene start (bp)  \\\n",
       " 0      ENSG00000000003     TSPAN6                 False      100627108.0   \n",
       " 1      ENSG00000000005       TNMD                 False      100584936.0   \n",
       " 2      ENSG00000000419       DPM1                 False       50934867.0   \n",
       " 3      ENSG00000000457      SCYL3                 False              NaN   \n",
       " 4      ENSG00000000460      FIRRM                 False              NaN   \n",
       " ...                ...        ...                   ...              ...   \n",
       " 41318  ENSG00000181404     WASHC1                 False              NaN   \n",
       " 41319  ENSG00000276581  SPATA31A5                 False              NaN   \n",
       " 41320  ENSG00000278848   TP53TG3F                 False              NaN   \n",
       " 41321  ENSG00000279782    PPIAL4F                 False              NaN   \n",
       " 41322  ENSG00000280267   PRAMEF26                 False              NaN   \n",
       " \n",
       "        Gene end (bp) Chromosome/scaffold name Protein stable ID  \n",
       " 0        100639991.0                        X   ENSP00000362111  \n",
       " 1        100599885.0                        X   ENSP00000362122  \n",
       " 2         50959140.0                       20   ENSP00000507119  \n",
       " 3                NaN                      NaN               NaN  \n",
       " 4                NaN                      NaN               NaN  \n",
       " ...              ...                      ...               ...  \n",
       " 41318            NaN                      NaN               NaN  \n",
       " 41319            NaN                      NaN               NaN  \n",
       " 41320            NaN                      NaN               NaN  \n",
       " 41321            NaN                      NaN               NaN  \n",
       " 41322            NaN                      NaN               NaN  \n",
       " \n",
       " [41323 rows x 7 columns]}"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "HWG"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "d19f71d9-8889-4663-b170-f16c74033756",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<1x1 sparse matrix of type '<class 'numpy.int64'>'\n",
       "\twith 0 stored elements in Compressed Sparse Row format>"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "HWG['A']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "4179f514-53f6-4de6-b878-f2df5a10a122",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    Missing TF: ENSG00000277462\n",
      "    Missing TF: ENSG00000214534\n",
      "    Missing TF: ENSG00000223614\n",
      "    Missing TF: ENSG00000228623\n",
      "    Missing TF: ENSG00000213793\n",
      "    Missing TF: DUX1_HUMAN\n",
      "    Missing TF: DUX3_HUMAN\n",
      "    Missing TF: ENSG00000260027\n",
      "    Missing TF: ENSG00000255192\n",
      "    Missing TF: ENSG00000230257\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_1965222/1148256144.py:35: DtypeWarning: Columns (32,34,38,40,50) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  HGNC_genes = pd.read_csv(HGNC_downloaded_data, sep='\\t')\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Stable ID</th>\n",
       "      <th>symbol</th>\n",
       "      <th>Transcription Factor</th>\n",
       "      <th>Gene start (bp)</th>\n",
       "      <th>Gene end (bp)</th>\n",
       "      <th>Chromosome/scaffold name</th>\n",
       "      <th>Protein stable ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>False</td>\n",
       "      <td>100627108.0</td>\n",
       "      <td>100639991.0</td>\n",
       "      <td>X</td>\n",
       "      <td>ENSP00000362111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000000005</td>\n",
       "      <td>TNMD</td>\n",
       "      <td>False</td>\n",
       "      <td>100584936.0</td>\n",
       "      <td>100599885.0</td>\n",
       "      <td>X</td>\n",
       "      <td>ENSP00000362122</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000000419</td>\n",
       "      <td>DPM1</td>\n",
       "      <td>False</td>\n",
       "      <td>50934867.0</td>\n",
       "      <td>50959140.0</td>\n",
       "      <td>20</td>\n",
       "      <td>ENSP00000507119</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000000457</td>\n",
       "      <td>SCYL3</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000000460</td>\n",
       "      <td>FIRRM</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41318</th>\n",
       "      <td>ENSG00000181404</td>\n",
       "      <td>WASHC1</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41319</th>\n",
       "      <td>ENSG00000276581</td>\n",
       "      <td>SPATA31A5</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41320</th>\n",
       "      <td>ENSG00000278848</td>\n",
       "      <td>TP53TG3F</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41321</th>\n",
       "      <td>ENSG00000279782</td>\n",
       "      <td>PPIAL4F</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41322</th>\n",
       "      <td>ENSG00000280267</td>\n",
       "      <td>PRAMEF26</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>41323 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Stable ID     symbol  Transcription Factor  Gene start (bp)  \\\n",
       "0      ENSG00000000003     TSPAN6                 False      100627108.0   \n",
       "1      ENSG00000000005       TNMD                 False      100584936.0   \n",
       "2      ENSG00000000419       DPM1                 False       50934867.0   \n",
       "3      ENSG00000000457      SCYL3                 False              NaN   \n",
       "4      ENSG00000000460      FIRRM                 False              NaN   \n",
       "...                ...        ...                   ...              ...   \n",
       "41318  ENSG00000181404     WASHC1                 False              NaN   \n",
       "41319  ENSG00000276581  SPATA31A5                 False              NaN   \n",
       "41320  ENSG00000278848   TP53TG3F                 False              NaN   \n",
       "41321  ENSG00000279782    PPIAL4F                 False              NaN   \n",
       "41322  ENSG00000280267   PRAMEF26                 False              NaN   \n",
       "\n",
       "       Gene end (bp) Chromosome/scaffold name Protein stable ID  \n",
       "0        100639991.0                        X   ENSP00000362111  \n",
       "1        100599885.0                        X   ENSP00000362122  \n",
       "2         50959140.0                       20   ENSP00000507119  \n",
       "3                NaN                      NaN               NaN  \n",
       "4                NaN                      NaN               NaN  \n",
       "...              ...                      ...               ...  \n",
       "41318            NaN                      NaN               NaN  \n",
       "41319            NaN                      NaN               NaN  \n",
       "41320            NaN                      NaN               NaN  \n",
       "41321            NaN                      NaN               NaN  \n",
       "41322            NaN                      NaN               NaN  \n",
       "\n",
       "[41323 rows x 7 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_path = load_data_path()\n",
    "\n",
    "# Get unique genes used in the HWG\n",
    "gene_ids, _ = stable_ID_map(0, 0, True)\n",
    "gene_ids = pd.unique(gene_ids)\n",
    "\n",
    "# Create index table\n",
    "indexTable = pd.DataFrame({\n",
    "    'Stable ID': gene_ids,\n",
    "    'Transcription Factor': [False] * len(gene_ids)\n",
    "})\n",
    "\n",
    "# Fill in transcription factors\n",
    "HuTF_downloaded_data =os.path.join(data_path,\"Downloaded/HuTF/HuTF_db.csv\")\n",
    "HuTF_db = pd.read_csv(HuTF_downloaded_data)\n",
    "isTF = HuTF_db['Is TF?']\n",
    "gene_stable_ids = HuTF_db['Ensembl ID']\n",
    "tf_ids = gene_stable_ids[isTF == \"Yes\"]\n",
    "for id in tf_ids:\n",
    "    idx = indexTable[indexTable['Stable ID'] == id].index\n",
    "    if idx.empty:\n",
    "        print(f\"    Missing TF: {id}\")\n",
    "    else:\n",
    "        indexTable.loc[idx, 'Transcription Factor'] = True\n",
    "        \n",
    "# Add in gene names where known\n",
    "# ENSEMBLE DATA\n",
    "GRCh_downloaded_data =os.path.join(data_path, \"Downloaded/Ensemble/GRCh38.p13.tsv\")\n",
    "GRCh_genes = pd.read_csv(GRCh_downloaded_data, sep='\\t')\n",
    "GRCh_genes['Gene stable ID'] = GRCh_genes['Gene stable ID'].str.strip()\n",
    "GRCh_genes['Gene name'] = GRCh_genes['Gene name'].str.strip()\n",
    "\n",
    "# HGNC DATA\n",
    "HGNC_downloaded_data = os.path.join(data_path, \"Downloaded/HGNC/gene_lookup_dictionary.tsv\")\n",
    "HGNC_genes = pd.read_csv(HGNC_downloaded_data, sep='\\t')\n",
    "HGNC_genes['ensembl_gene_id'] = HGNC_genes['ensembl_gene_id'].str.strip()\n",
    "HGNC_genes['symbol'] = HGNC_genes['symbol'].str.strip()\n",
    "\n",
    "# Merge in GRCH\n",
    "indexTable = pd.merge(indexTable, GRCh_genes, how='left',\n",
    "               left_on='Stable ID',\n",
    "               right_on='Gene stable ID',\n",
    "              )\n",
    "indexTable = indexTable[indexTable['Stable ID'].notna()]\n",
    "indexTable = indexTable.drop_duplicates(subset=['Stable ID'], keep='first')\n",
    "indexTable = indexTable.reset_index()\n",
    "indexTable = indexTable.drop('index', axis=1)\n",
    "\n",
    "# Merge in HGNC\n",
    "indexTable = pd.merge(indexTable, HGNC_genes, how='left',\n",
    "               left_on='Stable ID',\n",
    "               right_on='ensembl_gene_id',\n",
    "              )\n",
    "indexTable = indexTable[indexTable['Stable ID'].notna()]\n",
    "indexTable = indexTable.drop_duplicates(subset=['Stable ID'], keep='first')\n",
    "indexTable = indexTable.reset_index()\n",
    "indexTable = indexTable.drop('index', axis=1)\n",
    "\n",
    "# Reduce the table\n",
    "columns2keep = ['Stable ID', 'symbol', 'Transcription Factor', 'Gene start (bp)', 'Gene end (bp)', 'Chromosome/scaffold name', 'Protein stable ID']\n",
    "indexTable = indexTable[columns2keep]\n",
    "indexTable"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "31243723-31cb-4bab-a0f8-24dd56fc082a",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_1965222/2151804172.py:10: DtypeWarning: Columns (32,34,38,40,50) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  HGNC_genes = pd.read_csv(HGNC_downloaded_data, sep='\\t')\n"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "5875577f-d7a9-4761-aa4c-e22bb314f02a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Stable ID</th>\n",
       "      <th>Transcription Factor</th>\n",
       "      <th>Gene stable ID</th>\n",
       "      <th>Protein stable ID</th>\n",
       "      <th>Gene start (bp)</th>\n",
       "      <th>Gene end (bp)</th>\n",
       "      <th>Chromosome/scaffold name</th>\n",
       "      <th>Gene name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>ENSP00000362111</td>\n",
       "      <td>100627108.0</td>\n",
       "      <td>100639991.0</td>\n",
       "      <td>X</td>\n",
       "      <td>TSPAN6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000000005</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSG00000000005</td>\n",
       "      <td>ENSP00000362122</td>\n",
       "      <td>100584936.0</td>\n",
       "      <td>100599885.0</td>\n",
       "      <td>X</td>\n",
       "      <td>TNMD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000000419</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSG00000000419</td>\n",
       "      <td>ENSP00000507119</td>\n",
       "      <td>50934867.0</td>\n",
       "      <td>50959140.0</td>\n",
       "      <td>20</td>\n",
       "      <td>DPM1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000000457</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000000460</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41318</th>\n",
       "      <td>ENSG00000181404</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41319</th>\n",
       "      <td>ENSG00000276581</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41320</th>\n",
       "      <td>ENSG00000278848</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41321</th>\n",
       "      <td>ENSG00000279782</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41322</th>\n",
       "      <td>ENSG00000280267</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>41323 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Stable ID  Transcription Factor   Gene stable ID  \\\n",
       "0      ENSG00000000003                 False  ENSG00000000003   \n",
       "1      ENSG00000000005                 False  ENSG00000000005   \n",
       "2      ENSG00000000419                 False  ENSG00000000419   \n",
       "3      ENSG00000000457                 False              NaN   \n",
       "4      ENSG00000000460                 False              NaN   \n",
       "...                ...                   ...              ...   \n",
       "41318  ENSG00000181404                 False              NaN   \n",
       "41319  ENSG00000276581                 False              NaN   \n",
       "41320  ENSG00000278848                 False              NaN   \n",
       "41321  ENSG00000279782                 False              NaN   \n",
       "41322  ENSG00000280267                 False              NaN   \n",
       "\n",
       "      Protein stable ID  Gene start (bp)  Gene end (bp)  \\\n",
       "0       ENSP00000362111      100627108.0    100639991.0   \n",
       "1       ENSP00000362122      100584936.0    100599885.0   \n",
       "2       ENSP00000507119       50934867.0     50959140.0   \n",
       "3                   NaN              NaN            NaN   \n",
       "4                   NaN              NaN            NaN   \n",
       "...                 ...              ...            ...   \n",
       "41318               NaN              NaN            NaN   \n",
       "41319               NaN              NaN            NaN   \n",
       "41320               NaN              NaN            NaN   \n",
       "41321               NaN              NaN            NaN   \n",
       "41322               NaN              NaN            NaN   \n",
       "\n",
       "      Chromosome/scaffold name Gene name  \n",
       "0                            X    TSPAN6  \n",
       "1                            X      TNMD  \n",
       "2                           20      DPM1  \n",
       "3                          NaN       NaN  \n",
       "4                          NaN       NaN  \n",
       "...                        ...       ...  \n",
       "41318                      NaN       NaN  \n",
       "41319                      NaN       NaN  \n",
       "41320                      NaN       NaN  \n",
       "41321                      NaN       NaN  \n",
       "41322                      NaN       NaN  \n",
       "\n",
       "[41323 rows x 8 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "5e721968-6be7-4c31-9908-7e3a3d9cf9fa",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "9fd7a318-3693-459c-9aee-be9a9fb48678",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Stable ID</th>\n",
       "      <th>symbol</th>\n",
       "      <th>Transcription Factor</th>\n",
       "      <th>Gene start (bp)</th>\n",
       "      <th>Gene end (bp)</th>\n",
       "      <th>Chromosome/scaffold name</th>\n",
       "      <th>Protein stable ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>False</td>\n",
       "      <td>100627108.0</td>\n",
       "      <td>100639991.0</td>\n",
       "      <td>X</td>\n",
       "      <td>ENSP00000362111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000000005</td>\n",
       "      <td>TNMD</td>\n",
       "      <td>False</td>\n",
       "      <td>100584936.0</td>\n",
       "      <td>100599885.0</td>\n",
       "      <td>X</td>\n",
       "      <td>ENSP00000362122</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000000419</td>\n",
       "      <td>DPM1</td>\n",
       "      <td>False</td>\n",
       "      <td>50934867.0</td>\n",
       "      <td>50959140.0</td>\n",
       "      <td>20</td>\n",
       "      <td>ENSP00000507119</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000000457</td>\n",
       "      <td>SCYL3</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000000460</td>\n",
       "      <td>FIRRM</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41318</th>\n",
       "      <td>ENSG00000181404</td>\n",
       "      <td>WASHC1</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41319</th>\n",
       "      <td>ENSG00000276581</td>\n",
       "      <td>SPATA31A5</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41320</th>\n",
       "      <td>ENSG00000278848</td>\n",
       "      <td>TP53TG3F</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41321</th>\n",
       "      <td>ENSG00000279782</td>\n",
       "      <td>PPIAL4F</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41322</th>\n",
       "      <td>ENSG00000280267</td>\n",
       "      <td>PRAMEF26</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>41323 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Stable ID     symbol  Transcription Factor  Gene start (bp)  \\\n",
       "0      ENSG00000000003     TSPAN6                 False      100627108.0   \n",
       "1      ENSG00000000005       TNMD                 False      100584936.0   \n",
       "2      ENSG00000000419       DPM1                 False       50934867.0   \n",
       "3      ENSG00000000457      SCYL3                 False              NaN   \n",
       "4      ENSG00000000460      FIRRM                 False              NaN   \n",
       "...                ...        ...                   ...              ...   \n",
       "41318  ENSG00000181404     WASHC1                 False              NaN   \n",
       "41319  ENSG00000276581  SPATA31A5                 False              NaN   \n",
       "41320  ENSG00000278848   TP53TG3F                 False              NaN   \n",
       "41321  ENSG00000279782    PPIAL4F                 False              NaN   \n",
       "41322  ENSG00000280267   PRAMEF26                 False              NaN   \n",
       "\n",
       "       Gene end (bp) Chromosome/scaffold name Protein stable ID  \n",
       "0        100639991.0                        X   ENSP00000362111  \n",
       "1        100599885.0                        X   ENSP00000362122  \n",
       "2         50959140.0                       20   ENSP00000507119  \n",
       "3                NaN                      NaN               NaN  \n",
       "4                NaN                      NaN               NaN  \n",
       "...              ...                      ...               ...  \n",
       "41318            NaN                      NaN               NaN  \n",
       "41319            NaN                      NaN               NaN  \n",
       "41320            NaN                      NaN               NaN  \n",
       "41321            NaN                      NaN               NaN  \n",
       "41322            NaN                      NaN               NaN  \n",
       "\n",
       "[41323 rows x 7 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "columns2keep = ['Stable ID', 'symbol', 'Transcription Factor', 'Gene start (bp)', 'Gene end (bp)', 'Chromosome/scaffold name', 'Protein stable ID']\n",
    "indexTable = indexTable[columns2keep]\n",
    "indexTable"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "51e08d3a-4d26-4563-b6e3-f819a8373780",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'CDK1' in list(indexTable['symbol'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "b96e17e1-65dd-4212-8a71-86370f33f2ee",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1629"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "indexTable['Transcription Factor'].sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "032a5ac6-ab0d-4291-8f1b-dcf9549518ae",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Stable ID', 'Transcription Factor', 'Gene stable ID',\n",
       "       'Protein stable ID', 'Gene start (bp)', 'Gene end (bp)',\n",
       "       'Chromosome/scaffold name', 'Gene name', 'hgnc_id', 'symbol', 'name',\n",
       "       'locus_group', 'locus_type', 'status', 'location', 'location_sortable',\n",
       "       'alias_symbol', 'alias_name', 'prev_symbol', 'prev_name', 'gene_group',\n",
       "       'gene_group_id', 'date_approved_reserved', 'date_symbol_changed',\n",
       "       'date_name_changed', 'date_modified', 'entrez_id', 'ensembl_gene_id',\n",
       "       'vega_id', 'ucsc_id', 'ena', 'refseq_accession', 'ccds_id',\n",
       "       'uniprot_ids', 'pubmed_id', 'mgd_id', 'rgd_id', 'lsdb', 'cosmic',\n",
       "       'omim_id', 'mirbase', 'homeodb', 'snornabase', 'bioparadigms_slc',\n",
       "       'orphanet', 'pseudogene.org', 'horde_id', 'merops', 'imgt', 'iuphar',\n",
       "       'kznf_gene_catalog', 'mamit-trnadb', 'cd', 'lncrnadb', 'enzyme_id',\n",
       "       'intermediate_filament_db', 'rna_central_ids', 'lncipedia', 'gtrnadb',\n",
       "       'agr', 'mane_select', 'gencc'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "indexTable.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "bdf09bc2-eecc-4821-bdf5-35e6e9ba9ab5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene stable ID</th>\n",
       "      <th>Protein stable ID</th>\n",
       "      <th>Gene start (bp)</th>\n",
       "      <th>Gene end (bp)</th>\n",
       "      <th>Chromosome/scaffold name</th>\n",
       "      <th>Gene name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000210049</td>\n",
       "      <td>NaN</td>\n",
       "      <td>577</td>\n",
       "      <td>647</td>\n",
       "      <td>MT</td>\n",
       "      <td>MT-TF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000211459</td>\n",
       "      <td>NaN</td>\n",
       "      <td>648</td>\n",
       "      <td>1601</td>\n",
       "      <td>MT</td>\n",
       "      <td>MT-RNR1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000210077</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1602</td>\n",
       "      <td>1670</td>\n",
       "      <td>MT</td>\n",
       "      <td>MT-TV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000210082</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1671</td>\n",
       "      <td>3229</td>\n",
       "      <td>MT</td>\n",
       "      <td>MT-RNR2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000209082</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3230</td>\n",
       "      <td>3304</td>\n",
       "      <td>MT</td>\n",
       "      <td>MT-TL1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76116</th>\n",
       "      <td>ENSG00000131711</td>\n",
       "      <td>NaN</td>\n",
       "      <td>72107234</td>\n",
       "      <td>72209565</td>\n",
       "      <td>5</td>\n",
       "      <td>MAP1B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76117</th>\n",
       "      <td>ENSG00000131711</td>\n",
       "      <td>ENSP00000427194</td>\n",
       "      <td>72107234</td>\n",
       "      <td>72209565</td>\n",
       "      <td>5</td>\n",
       "      <td>MAP1B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76118</th>\n",
       "      <td>ENSG00000131711</td>\n",
       "      <td>ENSP00000423444</td>\n",
       "      <td>72107234</td>\n",
       "      <td>72209565</td>\n",
       "      <td>5</td>\n",
       "      <td>MAP1B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76119</th>\n",
       "      <td>ENSG00000131711</td>\n",
       "      <td>ENSP00000423416</td>\n",
       "      <td>72107234</td>\n",
       "      <td>72209565</td>\n",
       "      <td>5</td>\n",
       "      <td>MAP1B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76120</th>\n",
       "      <td>ENSG00000200637</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6017723</td>\n",
       "      <td>6017839</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>76121 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Gene stable ID Protein stable ID  Gene start (bp)  Gene end (bp)  \\\n",
       "0      ENSG00000210049               NaN              577            647   \n",
       "1      ENSG00000211459               NaN              648           1601   \n",
       "2      ENSG00000210077               NaN             1602           1670   \n",
       "3      ENSG00000210082               NaN             1671           3229   \n",
       "4      ENSG00000209082               NaN             3230           3304   \n",
       "...                ...               ...              ...            ...   \n",
       "76116  ENSG00000131711               NaN         72107234       72209565   \n",
       "76117  ENSG00000131711   ENSP00000427194         72107234       72209565   \n",
       "76118  ENSG00000131711   ENSP00000423444         72107234       72209565   \n",
       "76119  ENSG00000131711   ENSP00000423416         72107234       72209565   \n",
       "76120  ENSG00000200637               NaN          6017723        6017839   \n",
       "\n",
       "      Chromosome/scaffold name Gene name  \n",
       "0                           MT     MT-TF  \n",
       "1                           MT   MT-RNR1  \n",
       "2                           MT     MT-TV  \n",
       "3                           MT   MT-RNR2  \n",
       "4                           MT    MT-TL1  \n",
       "...                        ...       ...  \n",
       "76116                        5     MAP1B  \n",
       "76117                        5     MAP1B  \n",
       "76118                        5     MAP1B  \n",
       "76119                        5     MAP1B  \n",
       "76120                      NaN       NaN  \n",
       "\n",
       "[76121 rows x 6 columns]"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "GRCh_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "6fb54a58-d7ef-42b5-b384-38073f7feee0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Stable ID</th>\n",
       "      <th>Transcription Factor</th>\n",
       "      <th>Gene stable ID</th>\n",
       "      <th>Protein stable ID</th>\n",
       "      <th>Gene start (bp)</th>\n",
       "      <th>Gene end (bp)</th>\n",
       "      <th>Chromosome/scaffold name</th>\n",
       "      <th>Gene name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>ENSP00000362111</td>\n",
       "      <td>100627108.0</td>\n",
       "      <td>100639991.0</td>\n",
       "      <td>X</td>\n",
       "      <td>TSPAN6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000000005</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSG00000000005</td>\n",
       "      <td>ENSP00000362122</td>\n",
       "      <td>100584936.0</td>\n",
       "      <td>100599885.0</td>\n",
       "      <td>X</td>\n",
       "      <td>TNMD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000000419</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSG00000000419</td>\n",
       "      <td>ENSP00000507119</td>\n",
       "      <td>50934867.0</td>\n",
       "      <td>50959140.0</td>\n",
       "      <td>20</td>\n",
       "      <td>DPM1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000000457</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000000460</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41318</th>\n",
       "      <td>ENSG00000181404</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41319</th>\n",
       "      <td>ENSG00000276581</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41320</th>\n",
       "      <td>ENSG00000278848</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41321</th>\n",
       "      <td>ENSG00000279782</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41322</th>\n",
       "      <td>ENSG00000280267</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>41323 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Stable ID  Transcription Factor   Gene stable ID  \\\n",
       "0      ENSG00000000003                 False  ENSG00000000003   \n",
       "1      ENSG00000000005                 False  ENSG00000000005   \n",
       "2      ENSG00000000419                 False  ENSG00000000419   \n",
       "3      ENSG00000000457                 False              NaN   \n",
       "4      ENSG00000000460                 False              NaN   \n",
       "...                ...                   ...              ...   \n",
       "41318  ENSG00000181404                 False              NaN   \n",
       "41319  ENSG00000276581                 False              NaN   \n",
       "41320  ENSG00000278848                 False              NaN   \n",
       "41321  ENSG00000279782                 False              NaN   \n",
       "41322  ENSG00000280267                 False              NaN   \n",
       "\n",
       "      Protein stable ID  Gene start (bp)  Gene end (bp)  \\\n",
       "0       ENSP00000362111      100627108.0    100639991.0   \n",
       "1       ENSP00000362122      100584936.0    100599885.0   \n",
       "2       ENSP00000507119       50934867.0     50959140.0   \n",
       "3                   NaN              NaN            NaN   \n",
       "4                   NaN              NaN            NaN   \n",
       "...                 ...              ...            ...   \n",
       "41318               NaN              NaN            NaN   \n",
       "41319               NaN              NaN            NaN   \n",
       "41320               NaN              NaN            NaN   \n",
       "41321               NaN              NaN            NaN   \n",
       "41322               NaN              NaN            NaN   \n",
       "\n",
       "      Chromosome/scaffold name Gene name  \n",
       "0                            X    TSPAN6  \n",
       "1                            X      TNMD  \n",
       "2                           20      DPM1  \n",
       "3                          NaN       NaN  \n",
       "4                          NaN       NaN  \n",
       "...                        ...       ...  \n",
       "41318                      NaN       NaN  \n",
       "41319                      NaN       NaN  \n",
       "41320                      NaN       NaN  \n",
       "41321                      NaN       NaN  \n",
       "41322                      NaN       NaN  \n",
       "\n",
       "[41323 rows x 8 columns]"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "indexTable"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "8acfdb27-8481-4f43-be6d-54b21a8af0bc",
   "metadata": {},
   "outputs": [],
   "source": [
    "stable_IDs = pd.concat([GRCh_genes['Gene stable ID'], HGNC_genes['ensembl_gene_id']])\n",
    "gene_names = pd.concat([GRCh_genes['Gene name'], HGNC_genes['symbol']])\n",
    "stable_IDs = stable_IDs.astype(str)\n",
    "gene_names = gene_names.astype(str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "b99b2170-1874-486c-b55a-4a0d15cb4f1c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GeneID</th>\n",
       "      <th>Gene Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000210049</td>\n",
       "      <td>MT-TF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000211459</td>\n",
       "      <td>MT-RNR1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000210077</td>\n",
       "      <td>MT-TV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000210082</td>\n",
       "      <td>MT-RNR2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000209082</td>\n",
       "      <td>MT-TL1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43846</th>\n",
       "      <td>ENSG00000162378</td>\n",
       "      <td>ZYG11B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43847</th>\n",
       "      <td>ENSG00000159840</td>\n",
       "      <td>ZYX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43848</th>\n",
       "      <td>ENSG00000274572</td>\n",
       "      <td>ZYXP1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43849</th>\n",
       "      <td>ENSG00000074755</td>\n",
       "      <td>ZZEF1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43850</th>\n",
       "      <td>ENSG00000036549</td>\n",
       "      <td>ZZZ3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>119972 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                GeneID Gene Name\n",
       "0      ENSG00000210049     MT-TF\n",
       "1      ENSG00000211459   MT-RNR1\n",
       "2      ENSG00000210077     MT-TV\n",
       "3      ENSG00000210082   MT-RNR2\n",
       "4      ENSG00000209082    MT-TL1\n",
       "...                ...       ...\n",
       "43846  ENSG00000162378    ZYG11B\n",
       "43847  ENSG00000159840       ZYX\n",
       "43848  ENSG00000274572     ZYXP1\n",
       "43849  ENSG00000074755     ZZEF1\n",
       "43850  ENSG00000036549      ZZZ3\n",
       "\n",
       "[119972 rows x 2 columns]"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "idNameDf = pd.concat([stable_IDs, gene_names], axis=1)\n",
    "idNameDf.columns = ['GeneID', 'Gene Name']\n",
    "idNameDf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "8d9a9265-6b97-4619-99dd-1c2bde1fc62a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Stable ID</th>\n",
       "      <th>Transcription Factor</th>\n",
       "      <th>Gene stable ID</th>\n",
       "      <th>Protein stable ID</th>\n",
       "      <th>Gene start (bp)</th>\n",
       "      <th>Gene end (bp)</th>\n",
       "      <th>Chromosome/scaffold name</th>\n",
       "      <th>Gene name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>ENSP00000362111</td>\n",
       "      <td>100627108.0</td>\n",
       "      <td>100639991.0</td>\n",
       "      <td>X</td>\n",
       "      <td>TSPAN6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000000005</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSG00000000005</td>\n",
       "      <td>ENSP00000362122</td>\n",
       "      <td>100584936.0</td>\n",
       "      <td>100599885.0</td>\n",
       "      <td>X</td>\n",
       "      <td>TNMD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000000419</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSG00000000419</td>\n",
       "      <td>ENSP00000507119</td>\n",
       "      <td>50934867.0</td>\n",
       "      <td>50959140.0</td>\n",
       "      <td>20</td>\n",
       "      <td>DPM1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000000457</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000000460</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41318</th>\n",
       "      <td>ENSG00000181404</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41319</th>\n",
       "      <td>ENSG00000276581</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41320</th>\n",
       "      <td>ENSG00000278848</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41321</th>\n",
       "      <td>ENSG00000279782</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41322</th>\n",
       "      <td>ENSG00000280267</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>41323 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Stable ID  Transcription Factor   Gene stable ID  \\\n",
       "0      ENSG00000000003                 False  ENSG00000000003   \n",
       "1      ENSG00000000005                 False  ENSG00000000005   \n",
       "2      ENSG00000000419                 False  ENSG00000000419   \n",
       "3      ENSG00000000457                 False              NaN   \n",
       "4      ENSG00000000460                 False              NaN   \n",
       "...                ...                   ...              ...   \n",
       "41318  ENSG00000181404                 False              NaN   \n",
       "41319  ENSG00000276581                 False              NaN   \n",
       "41320  ENSG00000278848                 False              NaN   \n",
       "41321  ENSG00000279782                 False              NaN   \n",
       "41322  ENSG00000280267                 False              NaN   \n",
       "\n",
       "      Protein stable ID  Gene start (bp)  Gene end (bp)  \\\n",
       "0       ENSP00000362111      100627108.0    100639991.0   \n",
       "1       ENSP00000362122      100584936.0    100599885.0   \n",
       "2       ENSP00000507119       50934867.0     50959140.0   \n",
       "3                   NaN              NaN            NaN   \n",
       "4                   NaN              NaN            NaN   \n",
       "...                 ...              ...            ...   \n",
       "41318               NaN              NaN            NaN   \n",
       "41319               NaN              NaN            NaN   \n",
       "41320               NaN              NaN            NaN   \n",
       "41321               NaN              NaN            NaN   \n",
       "41322               NaN              NaN            NaN   \n",
       "\n",
       "      Chromosome/scaffold name Gene name  \n",
       "0                            X    TSPAN6  \n",
       "1                            X      TNMD  \n",
       "2                           20      DPM1  \n",
       "3                          NaN       NaN  \n",
       "4                          NaN       NaN  \n",
       "...                        ...       ...  \n",
       "41318                      NaN       NaN  \n",
       "41319                      NaN       NaN  \n",
       "41320                      NaN       NaN  \n",
       "41321                      NaN       NaN  \n",
       "41322                      NaN       NaN  \n",
       "\n",
       "[41323 rows x 8 columns]"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "indexTable"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "da208a8d-4e06-49cc-85c6-b72768b2eeca",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "KeyboardInterrupt\n",
      "\n"
     ]
    }
   ],
   "source": [
    "id_to_name_mapping = {}\n",
    "for id in pd.unique(stable_IDs):\n",
    "    # Get the index of the current ID in indexTable\n",
    "    idx = indexTable[indexTable['Stable ID'] == id].index\n",
    "    if not idx.empty:\n",
    "        # Get the gene name corresponding to the current ID\n",
    "        gene_name = gene_names[stable_IDs == id].values[0]\n",
    "        # Update the dictionary\n",
    "        id_to_name_mapping[id] = gene_name\n",
    "\n",
    "indexTable['Gene Name'] = indexTable['Stable ID'].map(id_to_name_mapping).fillna('Unknown')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b21c1d8c-2c56-44bb-928a-45d0275f6c3f",
   "metadata": {},
   "outputs": [],
   "source": [
    "indexTable"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "70a3fd6f-1330-44b2-b4eb-f87068de280a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>hgnc_id</th>\n",
       "      <th>symbol</th>\n",
       "      <th>name</th>\n",
       "      <th>locus_group</th>\n",
       "      <th>locus_type</th>\n",
       "      <th>status</th>\n",
       "      <th>location</th>\n",
       "      <th>location_sortable</th>\n",
       "      <th>alias_symbol</th>\n",
       "      <th>alias_name</th>\n",
       "      <th>...</th>\n",
       "      <th>cd</th>\n",
       "      <th>lncrnadb</th>\n",
       "      <th>enzyme_id</th>\n",
       "      <th>intermediate_filament_db</th>\n",
       "      <th>rna_central_ids</th>\n",
       "      <th>lncipedia</th>\n",
       "      <th>gtrnadb</th>\n",
       "      <th>agr</th>\n",
       "      <th>mane_select</th>\n",
       "      <th>gencc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HGNC:5</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>alpha-1-B glycoprotein</td>\n",
       "      <td>protein-coding gene</td>\n",
       "      <td>gene with protein product</td>\n",
       "      <td>Approved</td>\n",
       "      <td>19q13.43</td>\n",
       "      <td>19q13.43</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:5</td>\n",
       "      <td>ENST00000263100.8|NM_130786.4</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HGNC:37133</td>\n",
       "      <td>A1BG-AS1</td>\n",
       "      <td>A1BG antisense RNA 1</td>\n",
       "      <td>non-coding RNA</td>\n",
       "      <td>RNA, long non-coding</td>\n",
       "      <td>Approved</td>\n",
       "      <td>19q13.43</td>\n",
       "      <td>19q13.43</td>\n",
       "      <td>FLJ23569</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A1BG-AS1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:37133</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HGNC:24086</td>\n",
       "      <td>A1CF</td>\n",
       "      <td>APOBEC1 complementation factor</td>\n",
       "      <td>protein-coding gene</td>\n",
       "      <td>gene with protein product</td>\n",
       "      <td>Approved</td>\n",
       "      <td>10q11.23</td>\n",
       "      <td>10q11.23</td>\n",
       "      <td>ACF|ASP|ACF64|ACF65|APOBEC1CF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:24086</td>\n",
       "      <td>ENST00000373997.8|NM_014576.4</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HGNC:7</td>\n",
       "      <td>A2M</td>\n",
       "      <td>alpha-2-macroglobulin</td>\n",
       "      <td>protein-coding gene</td>\n",
       "      <td>gene with protein product</td>\n",
       "      <td>Approved</td>\n",
       "      <td>12p13.31</td>\n",
       "      <td>12p13.31</td>\n",
       "      <td>FWP007|S863-7|CPAMD5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:7</td>\n",
       "      <td>ENST00000318602.12|NM_000014.6</td>\n",
       "      <td>HGNC:7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HGNC:27057</td>\n",
       "      <td>A2M-AS1</td>\n",
       "      <td>A2M antisense RNA 1</td>\n",
       "      <td>non-coding RNA</td>\n",
       "      <td>RNA, long non-coding</td>\n",
       "      <td>Approved</td>\n",
       "      <td>12p13.31</td>\n",
       "      <td>12p13.31</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A2M-AS1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:27057</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43846</th>\n",
       "      <td>HGNC:25820</td>\n",
       "      <td>ZYG11B</td>\n",
       "      <td>zyg-11 family member B, cell cycle regulator</td>\n",
       "      <td>protein-coding gene</td>\n",
       "      <td>gene with protein product</td>\n",
       "      <td>Approved</td>\n",
       "      <td>1p32.3</td>\n",
       "      <td>01p32.3</td>\n",
       "      <td>FLJ13456</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:25820</td>\n",
       "      <td>ENST00000294353.7|NM_024646.3</td>\n",
       "      <td>HGNC:25820</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43847</th>\n",
       "      <td>HGNC:13200</td>\n",
       "      <td>ZYX</td>\n",
       "      <td>zyxin</td>\n",
       "      <td>protein-coding gene</td>\n",
       "      <td>gene with protein product</td>\n",
       "      <td>Approved</td>\n",
       "      <td>7q34</td>\n",
       "      <td>07q34</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:13200</td>\n",
       "      <td>ENST00000322764.10|NM_003461.5</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43848</th>\n",
       "      <td>HGNC:51695</td>\n",
       "      <td>ZYXP1</td>\n",
       "      <td>zyxin pseudogene 1</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>Approved</td>\n",
       "      <td>8q24.23</td>\n",
       "      <td>08q24.23</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:51695</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43849</th>\n",
       "      <td>HGNC:29027</td>\n",
       "      <td>ZZEF1</td>\n",
       "      <td>zinc finger ZZ-type and EF-hand domain contain...</td>\n",
       "      <td>protein-coding gene</td>\n",
       "      <td>gene with protein product</td>\n",
       "      <td>Approved</td>\n",
       "      <td>17p13.2</td>\n",
       "      <td>17p13.2</td>\n",
       "      <td>KIAA0399|ZZZ4|FLJ10821</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:29027</td>\n",
       "      <td>ENST00000381638.7|NM_015113.4</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43850</th>\n",
       "      <td>HGNC:24523</td>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>zinc finger ZZ-type containing 3</td>\n",
       "      <td>protein-coding gene</td>\n",
       "      <td>gene with protein product</td>\n",
       "      <td>Approved</td>\n",
       "      <td>1p31.1</td>\n",
       "      <td>01p31.1</td>\n",
       "      <td>DKFZP564I052|ATAC1</td>\n",
       "      <td>ATAC component 1 homolog (Drosophila)</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:24523</td>\n",
       "      <td>ENST00000370801.8|NM_015534.6</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>43851 rows × 54 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          hgnc_id    symbol  \\\n",
       "0          HGNC:5      A1BG   \n",
       "1      HGNC:37133  A1BG-AS1   \n",
       "2      HGNC:24086      A1CF   \n",
       "3          HGNC:7       A2M   \n",
       "4      HGNC:27057   A2M-AS1   \n",
       "...           ...       ...   \n",
       "43846  HGNC:25820    ZYG11B   \n",
       "43847  HGNC:13200       ZYX   \n",
       "43848  HGNC:51695     ZYXP1   \n",
       "43849  HGNC:29027     ZZEF1   \n",
       "43850  HGNC:24523      ZZZ3   \n",
       "\n",
       "                                                    name          locus_group  \\\n",
       "0                                 alpha-1-B glycoprotein  protein-coding gene   \n",
       "1                                   A1BG antisense RNA 1       non-coding RNA   \n",
       "2                         APOBEC1 complementation factor  protein-coding gene   \n",
       "3                                  alpha-2-macroglobulin  protein-coding gene   \n",
       "4                                    A2M antisense RNA 1       non-coding RNA   \n",
       "...                                                  ...                  ...   \n",
       "43846       zyg-11 family member B, cell cycle regulator  protein-coding gene   \n",
       "43847                                              zyxin  protein-coding gene   \n",
       "43848                                 zyxin pseudogene 1           pseudogene   \n",
       "43849  zinc finger ZZ-type and EF-hand domain contain...  protein-coding gene   \n",
       "43850                   zinc finger ZZ-type containing 3  protein-coding gene   \n",
       "\n",
       "                      locus_type    status  location location_sortable  \\\n",
       "0      gene with protein product  Approved  19q13.43          19q13.43   \n",
       "1           RNA, long non-coding  Approved  19q13.43          19q13.43   \n",
       "2      gene with protein product  Approved  10q11.23          10q11.23   \n",
       "3      gene with protein product  Approved  12p13.31          12p13.31   \n",
       "4           RNA, long non-coding  Approved  12p13.31          12p13.31   \n",
       "...                          ...       ...       ...               ...   \n",
       "43846  gene with protein product  Approved    1p32.3           01p32.3   \n",
       "43847  gene with protein product  Approved      7q34             07q34   \n",
       "43848                 pseudogene  Approved   8q24.23          08q24.23   \n",
       "43849  gene with protein product  Approved   17p13.2           17p13.2   \n",
       "43850  gene with protein product  Approved    1p31.1           01p31.1   \n",
       "\n",
       "                        alias_symbol                             alias_name  \\\n",
       "0                                NaN                                    NaN   \n",
       "1                           FLJ23569                                    NaN   \n",
       "2      ACF|ASP|ACF64|ACF65|APOBEC1CF                                    NaN   \n",
       "3               FWP007|S863-7|CPAMD5                                    NaN   \n",
       "4                                NaN                                    NaN   \n",
       "...                              ...                                    ...   \n",
       "43846                       FLJ13456                                    NaN   \n",
       "43847                            NaN                                    NaN   \n",
       "43848                            NaN                                    NaN   \n",
       "43849         KIAA0399|ZZZ4|FLJ10821                                    NaN   \n",
       "43850             DKFZP564I052|ATAC1  ATAC component 1 homolog (Drosophila)   \n",
       "\n",
       "       ...   cd lncrnadb enzyme_id intermediate_filament_db rna_central_ids  \\\n",
       "0      ...  NaN      NaN       NaN                      NaN             NaN   \n",
       "1      ...  NaN      NaN       NaN                      NaN             NaN   \n",
       "2      ...  NaN      NaN       NaN                      NaN             NaN   \n",
       "3      ...  NaN      NaN       NaN                      NaN             NaN   \n",
       "4      ...  NaN      NaN       NaN                      NaN             NaN   \n",
       "...    ...  ...      ...       ...                      ...             ...   \n",
       "43846  ...  NaN      NaN       NaN                      NaN             NaN   \n",
       "43847  ...  NaN      NaN       NaN                      NaN             NaN   \n",
       "43848  ...  NaN      NaN       NaN                      NaN             NaN   \n",
       "43849  ...  NaN      NaN       NaN                      NaN             NaN   \n",
       "43850  ...  NaN      NaN       NaN                      NaN             NaN   \n",
       "\n",
       "      lncipedia gtrnadb         agr                     mane_select  \\\n",
       "0           NaN     NaN      HGNC:5   ENST00000263100.8|NM_130786.4   \n",
       "1      A1BG-AS1     NaN  HGNC:37133                             NaN   \n",
       "2           NaN     NaN  HGNC:24086   ENST00000373997.8|NM_014576.4   \n",
       "3           NaN     NaN      HGNC:7  ENST00000318602.12|NM_000014.6   \n",
       "4       A2M-AS1     NaN  HGNC:27057                             NaN   \n",
       "...         ...     ...         ...                             ...   \n",
       "43846       NaN     NaN  HGNC:25820   ENST00000294353.7|NM_024646.3   \n",
       "43847       NaN     NaN  HGNC:13200  ENST00000322764.10|NM_003461.5   \n",
       "43848       NaN     NaN  HGNC:51695                             NaN   \n",
       "43849       NaN     NaN  HGNC:29027   ENST00000381638.7|NM_015113.4   \n",
       "43850       NaN     NaN  HGNC:24523   ENST00000370801.8|NM_015534.6   \n",
       "\n",
       "            gencc  \n",
       "0             NaN  \n",
       "1             NaN  \n",
       "2             NaN  \n",
       "3          HGNC:7  \n",
       "4             NaN  \n",
       "...           ...  \n",
       "43846  HGNC:25820  \n",
       "43847         NaN  \n",
       "43848         NaN  \n",
       "43849         NaN  \n",
       "43850         NaN  \n",
       "\n",
       "[43851 rows x 54 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "HGNC_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "1276f436-4e26-4eb0-a2d2-6cc81e591e9c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene stable ID</th>\n",
       "      <th>Protein stable ID</th>\n",
       "      <th>Gene start (bp)</th>\n",
       "      <th>Gene end (bp)</th>\n",
       "      <th>Chromosome/scaffold name</th>\n",
       "      <th>Gene name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000210049</td>\n",
       "      <td>NaN</td>\n",
       "      <td>577</td>\n",
       "      <td>647</td>\n",
       "      <td>MT</td>\n",
       "      <td>MT-TF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000211459</td>\n",
       "      <td>NaN</td>\n",
       "      <td>648</td>\n",
       "      <td>1601</td>\n",
       "      <td>MT</td>\n",
       "      <td>MT-RNR1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000210077</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1602</td>\n",
       "      <td>1670</td>\n",
       "      <td>MT</td>\n",
       "      <td>MT-TV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000210082</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1671</td>\n",
       "      <td>3229</td>\n",
       "      <td>MT</td>\n",
       "      <td>MT-RNR2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000209082</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3230</td>\n",
       "      <td>3304</td>\n",
       "      <td>MT</td>\n",
       "      <td>MT-TL1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76116</th>\n",
       "      <td>ENSG00000131711</td>\n",
       "      <td>NaN</td>\n",
       "      <td>72107234</td>\n",
       "      <td>72209565</td>\n",
       "      <td>5</td>\n",
       "      <td>MAP1B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76117</th>\n",
       "      <td>ENSG00000131711</td>\n",
       "      <td>ENSP00000427194</td>\n",
       "      <td>72107234</td>\n",
       "      <td>72209565</td>\n",
       "      <td>5</td>\n",
       "      <td>MAP1B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76118</th>\n",
       "      <td>ENSG00000131711</td>\n",
       "      <td>ENSP00000423444</td>\n",
       "      <td>72107234</td>\n",
       "      <td>72209565</td>\n",
       "      <td>5</td>\n",
       "      <td>MAP1B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76119</th>\n",
       "      <td>ENSG00000131711</td>\n",
       "      <td>ENSP00000423416</td>\n",
       "      <td>72107234</td>\n",
       "      <td>72209565</td>\n",
       "      <td>5</td>\n",
       "      <td>MAP1B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76120</th>\n",
       "      <td>ENSG00000200637</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6017723</td>\n",
       "      <td>6017839</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>76121 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Gene stable ID Protein stable ID  Gene start (bp)  Gene end (bp)  \\\n",
       "0      ENSG00000210049               NaN              577            647   \n",
       "1      ENSG00000211459               NaN              648           1601   \n",
       "2      ENSG00000210077               NaN             1602           1670   \n",
       "3      ENSG00000210082               NaN             1671           3229   \n",
       "4      ENSG00000209082               NaN             3230           3304   \n",
       "...                ...               ...              ...            ...   \n",
       "76116  ENSG00000131711               NaN         72107234       72209565   \n",
       "76117  ENSG00000131711   ENSP00000427194         72107234       72209565   \n",
       "76118  ENSG00000131711   ENSP00000423444         72107234       72209565   \n",
       "76119  ENSG00000131711   ENSP00000423416         72107234       72209565   \n",
       "76120  ENSG00000200637               NaN          6017723        6017839   \n",
       "\n",
       "      Chromosome/scaffold name Gene name  \n",
       "0                           MT     MT-TF  \n",
       "1                           MT   MT-RNR1  \n",
       "2                           MT     MT-TV  \n",
       "3                           MT   MT-RNR2  \n",
       "4                           MT    MT-TL1  \n",
       "...                        ...       ...  \n",
       "76116                        5     MAP1B  \n",
       "76117                        5     MAP1B  \n",
       "76118                        5     MAP1B  \n",
       "76119                        5     MAP1B  \n",
       "76120                      NaN       NaN  \n",
       "\n",
       "[76121 rows x 6 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "GRCh_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "caca0c8e-b76d-4fd9-bb06-efa23917bb14",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Stable ID</th>\n",
       "      <th>Gene Name</th>\n",
       "      <th>Transcription Factor</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>None</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000000005</td>\n",
       "      <td>None</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000000419</td>\n",
       "      <td>None</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000000457</td>\n",
       "      <td>None</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000000460</td>\n",
       "      <td>None</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41318</th>\n",
       "      <td>ENSG00000181404</td>\n",
       "      <td>None</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41319</th>\n",
       "      <td>ENSG00000276581</td>\n",
       "      <td>None</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41320</th>\n",
       "      <td>ENSG00000278848</td>\n",
       "      <td>None</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41321</th>\n",
       "      <td>ENSG00000279782</td>\n",
       "      <td>None</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41322</th>\n",
       "      <td>ENSG00000280267</td>\n",
       "      <td>None</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>41323 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Stable ID Gene Name  Transcription Factor\n",
       "0      ENSG00000000003      None                 False\n",
       "1      ENSG00000000005      None                 False\n",
       "2      ENSG00000000419      None                 False\n",
       "3      ENSG00000000457      None                 False\n",
       "4      ENSG00000000460      None                 False\n",
       "...                ...       ...                   ...\n",
       "41318  ENSG00000181404      None                 False\n",
       "41319  ENSG00000276581      None                 False\n",
       "41320  ENSG00000278848      None                 False\n",
       "41321  ENSG00000279782      None                 False\n",
       "41322  ENSG00000280267      None                 False\n",
       "\n",
       "[41323 rows x 3 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "indexTable"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0b12d857-a6cc-4cc0-9e75-58b847e1259f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Combine ENSEMBLE and HGNC data\n",
    "stable_IDs = pd.concat([GRCh_genes['Gene stable ID'], HGNC_genes['ensembl_gene_id']])\n",
    "gene_names = pd.concat([GRCh_genes['Gene name'], HGNC_genes['symbol']])\n",
    "\n",
    "stable_IDs = stable_IDs.astype(str)\n",
    "gene_names = gene_names.astype(str)\n",
    "# for id in pd.unique(stable_IDs):\n",
    "#     idx = indexTable[indexTable['Stable ID'] == id].index\n",
    "#     if not idx.empty:\n",
    "#         indexTable.at[idx[0], 'Gene Name'] = gene_names[stable_IDs == id].values[0]\n",
    "# id_to_name_mapping = pd.Series(gene_names.values, index=stable_IDs).to_dict()\n",
    "id_to_name_mapping = {}\n",
    "for id in pd.unique(stable_IDs):\n",
    "    # Get the index of the current ID in indexTable\n",
    "    idx = indexTable[indexTable['Stable ID'] == id].index\n",
    "    if not idx.empty:\n",
    "        # Get the gene name corresponding to the current ID\n",
    "        gene_name = gene_names[stable_IDs == id].values[0]\n",
    "        # Update the dictionary\n",
    "        id_to_name_mapping[id] = gene_name\n",
    "\n",
    "indexTable['Gene Name'] = indexTable['Stable ID'].map(id_to_name_mapping).fillna('Unknown')\n",
    "\n",
    "# Add in gene locations/positional information\n",
    "mat_data_path = os.path.join(data_path, \"HWG nodes.mat\")\n",
    "mat_data = scipy.io.loadmat(mat_data_path)\n",
    "\n",
    "def extract_matlab_opaque(mat_obj):\n",
    "    result = {}\n",
    "    if isinstance(mat_obj, np.ndarray):\n",
    "        if mat_obj.dtype.names:\n",
    "            for name in mat_obj.dtype.names:\n",
    "                result[name] = extract_matlab_opaque(mat_obj[name])\n",
    "        else:\n",
    "            result = mat_obj.tolist()\n",
    "    elif isinstance(mat_obj, scipy.io.matlab.mio5_params.mat_struct):\n",
    "        for field_name in mat_obj._fieldnames:\n",
    "            result[field_name] = extract_matlab_opaque(getattr(mat_obj, field_name))\n",
    "    else:\n",
    "        result = mat_obj\n",
    "    return result\n",
    "\n",
    "extracted_data = extract_matlab_opaque(mat_data['None'])\n",
    "\n",
    "# Assuming 'cur_tbl' data is extracted correctly\n",
    "if 'cur_tbl' in extracted_data:\n",
    "    cur_tbl = extracted_data['cur_tbl']\n",
    "    cur_tbl_df = pd.DataFrame(cur_tbl)\n",
    "    \n",
    "    # Ensure correct column names (you may need to adjust based on actual data)\n",
    "    cur_tbl_df.columns = ['Chromosome', 'GeneStart', 'GeneEnd']\n",
    "\n",
    "    curtable_id2idx = {str(cur_tbl_df.index[i]): i for i in range(len(cur_tbl_df))}\n",
    "\n",
    "    gene_info = pd.DataFrame(columns=['Chromosome', 'Start', 'End'])\n",
    "    for i in range(len(indexTable)):\n",
    "        stable_id = indexTable.at[i, 'Stable ID']\n",
    "        if stable_id in curtable_id2idx:\n",
    "            idx = curtable_id2idx[stable_id]\n",
    "            chr = cur_tbl_df.at[idx, 'Chromosome']\n",
    "            start = cur_tbl_df.at[idx, 'GeneStart']\n",
    "            stop = cur_tbl_df.at[idx, 'GeneEnd']\n",
    "            gene_info = gene_info.append({'Chromosome': chr, 'Start': start, 'End': stop}, ignore_index=True)\n",
    "        else:\n",
    "            gene_info = gene_info.append({'Chromosome': 0, 'Start': 0, 'End': 0}, ignore_index=True)\n",
    "\n",
    "    indexTable = pd.concat([indexTable, gene_info], axis=1)\n",
    "\n",
    "return indexTable"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "f65aef75-ef1e-482d-bddc-0dde00a28d5f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    Missing TF: ENSG00000277462\n",
      "    Missing TF: ENSG00000214534\n",
      "    Missing TF: ENSG00000223614\n",
      "    Missing TF: ENSG00000228623\n",
      "    Missing TF: ENSG00000213793\n",
      "    Missing TF: DUX1_HUMAN\n",
      "    Missing TF: DUX3_HUMAN\n",
      "    Missing TF: ENSG00000260027\n",
      "    Missing TF: ENSG00000255192\n",
      "    Missing TF: ENSG00000230257\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jpic/HardwiredGenome-Python/HWG/build.py:242: DtypeWarning: Columns (32,34,38,40,50) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  HGNC_genes = pd.read_csv(HGNC_downloaded_data, sep='\\t')\n",
      "\n",
      "KeyboardInterrupt\n",
      "\n"
     ]
    }
   ],
   "source": [
    "build.buildIndexTable()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "326d7508-eaa0-4df9-b760-90f26bafb7c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_path = build.load_data_path()\n",
    "gene_ids, _ = build.stable_ID_map(0, 0, True)\n",
    "gene_ids = pd.unique(gene_ids)\n",
    "indexTable = pd.DataFrame({\n",
    "        'Stable ID': gene_ids,\n",
    "        'Gene Name': [None] * len(gene_ids),\n",
    "        'Transcription Factor': [False] * len(gene_ids)\n",
    "    })"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "1c656437-baa5-444f-b6ea-c3a9387e1815",
   "metadata": {},
   "outputs": [],
   "source": [
    "HuTF_downloaded_data =os.path.join(data_path,\"Downloaded/HuTF/HuTF_db.csv\")\n",
    "HuTF_db = pd.read_csv(HuTF_downloaded_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "6b1302ab-4342-4bb6-9334-84b7e3fccabb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    Missing TF: ENSG00000277462\n",
      "    Missing TF: ENSG00000214534\n",
      "    Missing TF: ENSG00000223614\n",
      "    Missing TF: ENSG00000228623\n",
      "    Missing TF: ENSG00000213793\n",
      "    Missing TF: DUX1_HUMAN\n",
      "    Missing TF: DUX3_HUMAN\n",
      "    Missing TF: ENSG00000260027\n",
      "    Missing TF: ENSG00000255192\n",
      "    Missing TF: ENSG00000230257\n"
     ]
    }
   ],
   "source": [
    "isTF = HuTF_db['Is TF?']\n",
    "gene_stable_ids = HuTF_db['Ensembl ID']\n",
    "\n",
    "tf_ids = gene_stable_ids[isTF == \"Yes\"]\n",
    "for id in tf_ids:\n",
    "    idx = indexTable[indexTable['Stable ID'] == id].index\n",
    "    if idx.empty:\n",
    "        print(f\"    Missing TF: {id}\")\n",
    "    else:\n",
    "        indexTable.loc[idx, 'Transcription Factor'] = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "ac5cb82d-4fcf-49d7-b5ea-2258bad3b8dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "GRCh_downloaded_data =os.path.join(data_path, \"Downloaded/Ensemble/GRCh38.p13.tsv\")\n",
    "GRCh_genes = pd.read_csv(GRCh_downloaded_data, sep='\\t')\n",
    "GRCh_genes['Gene stable ID'] = GRCh_genes['Gene stable ID'].str.strip()\n",
    "GRCh_genes['Gene name'] = GRCh_genes['Gene name'].str.strip()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "e1a321e4-0b0a-4b24-9e94-f7c9d92c52b6",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_2614989/821611533.py:2: DtypeWarning: Columns (32,34,38,40,50) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  HGNC_genes = pd.read_csv(HGNC_downloaded_data, sep='\\t')\n"
     ]
    }
   ],
   "source": [
    "HGNC_downloaded_data = os.path.join(data_path, \"Downloaded/HGNC/gene_lookup_dictionary.tsv\")\n",
    "HGNC_genes = pd.read_csv(HGNC_downloaded_data, sep='\\t')\n",
    "HGNC_genes['ensembl_gene_id'] = HGNC_genes['ensembl_gene_id'].str.strip()\n",
    "HGNC_genes['symbol'] = HGNC_genes['symbol'].str.strip()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "c4ea7898-4c4f-45e5-936d-fffa41b16ee1",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "KeyboardInterrupt\n",
      "\n"
     ]
    }
   ],
   "source": [
    "stable_IDs = pd.concat([GRCh_genes['Gene stable ID'], HGNC_genes['ensembl_gene_id']])\n",
    "gene_names = pd.concat([GRCh_genes['Gene name'], HGNC_genes['symbol']])\n",
    "\n",
    "stable_IDs = stable_IDs.astype(str)\n",
    "gene_names = gene_names.astype(str)\n",
    "# for id in pd.unique(stable_IDs):\n",
    "#     idx = indexTable[indexTable['Stable ID'] == id].index\n",
    "#     if not idx.empty:\n",
    "#         indexTable.at[idx[0], 'Gene Name'] = gene_names[stable_IDs == id].values[0]\n",
    "# id_to_name_mapping = pd.Series(gene_names.values, index=stable_IDs).to_dict()\n",
    "id_to_name_mapping = {}\n",
    "for id in pd.unique(stable_IDs):\n",
    "    # Get the index of the current ID in indexTable\n",
    "    idx = indexTable[indexTable['Stable ID'] == id].index\n",
    "    if not idx.empty:\n",
    "        # Get the gene name corresponding to the current ID\n",
    "        gene_name = gene_names[stable_IDs == id].values[0]\n",
    "        # Update the dictionary\n",
    "        id_to_name_mapping[id] = gene_name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "67d9f532-2b70-4d0a-a092-cd2a286f31c7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    Missing TF: ENSG00000277462\n",
      "    Missing TF: ENSG00000214534\n",
      "    Missing TF: ENSG00000223614\n",
      "    Missing TF: ENSG00000228623\n",
      "    Missing TF: ENSG00000213793\n",
      "    Missing TF: DUX1_HUMAN\n",
      "    Missing TF: DUX3_HUMAN\n",
      "    Missing TF: ENSG00000260027\n",
      "    Missing TF: ENSG00000255192\n",
      "    Missing TF: ENSG00000230257\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jpic/HardwiredGenome-Python/HWG/build.py:245: DtypeWarning: Columns (32,34,38,40,50) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  \n"
     ]
    }
   ],
   "source": [
    "indextable = build.buildIndexTable()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "39cf3ac9-b6c2-48b0-8e19-536df1a40268",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    HGNC Downloaded\n"
     ]
    }
   ],
   "source": [
    "build.download_HGNC()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "96067185-cefc-409d-ba46-5ac9187d62c7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene stable ID</th>\n",
       "      <th>Protein stable ID</th>\n",
       "      <th>Gene start (bp)</th>\n",
       "      <th>Gene end (bp)</th>\n",
       "      <th>Chromosome/scaffold name</th>\n",
       "      <th>Gene name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000210049</td>\n",
       "      <td>NaN</td>\n",
       "      <td>577</td>\n",
       "      <td>647</td>\n",
       "      <td>MT</td>\n",
       "      <td>MT-TF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000211459</td>\n",
       "      <td>NaN</td>\n",
       "      <td>648</td>\n",
       "      <td>1601</td>\n",
       "      <td>MT</td>\n",
       "      <td>MT-RNR1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000210077</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1602</td>\n",
       "      <td>1670</td>\n",
       "      <td>MT</td>\n",
       "      <td>MT-TV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000210082</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1671</td>\n",
       "      <td>3229</td>\n",
       "      <td>MT</td>\n",
       "      <td>MT-RNR2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000209082</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3230</td>\n",
       "      <td>3304</td>\n",
       "      <td>MT</td>\n",
       "      <td>MT-TL1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76116</th>\n",
       "      <td>ENSG00000131711</td>\n",
       "      <td>NaN</td>\n",
       "      <td>72107234</td>\n",
       "      <td>72209565</td>\n",
       "      <td>5</td>\n",
       "      <td>MAP1B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76117</th>\n",
       "      <td>ENSG00000131711</td>\n",
       "      <td>ENSP00000427194</td>\n",
       "      <td>72107234</td>\n",
       "      <td>72209565</td>\n",
       "      <td>5</td>\n",
       "      <td>MAP1B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76118</th>\n",
       "      <td>ENSG00000131711</td>\n",
       "      <td>ENSP00000423444</td>\n",
       "      <td>72107234</td>\n",
       "      <td>72209565</td>\n",
       "      <td>5</td>\n",
       "      <td>MAP1B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76119</th>\n",
       "      <td>ENSG00000131711</td>\n",
       "      <td>ENSP00000423416</td>\n",
       "      <td>72107234</td>\n",
       "      <td>72209565</td>\n",
       "      <td>5</td>\n",
       "      <td>MAP1B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76120</th>\n",
       "      <td>ENSG00000200637</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6017723</td>\n",
       "      <td>6017839</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>76121 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Gene stable ID Protein stable ID  Gene start (bp)  Gene end (bp)  \\\n",
       "0      ENSG00000210049               NaN              577            647   \n",
       "1      ENSG00000211459               NaN              648           1601   \n",
       "2      ENSG00000210077               NaN             1602           1670   \n",
       "3      ENSG00000210082               NaN             1671           3229   \n",
       "4      ENSG00000209082               NaN             3230           3304   \n",
       "...                ...               ...              ...            ...   \n",
       "76116  ENSG00000131711               NaN         72107234       72209565   \n",
       "76117  ENSG00000131711   ENSP00000427194         72107234       72209565   \n",
       "76118  ENSG00000131711   ENSP00000423444         72107234       72209565   \n",
       "76119  ENSG00000131711   ENSP00000423416         72107234       72209565   \n",
       "76120  ENSG00000200637               NaN          6017723        6017839   \n",
       "\n",
       "      Chromosome/scaffold name Gene name  \n",
       "0                           MT     MT-TF  \n",
       "1                           MT   MT-RNR1  \n",
       "2                           MT     MT-TV  \n",
       "3                           MT   MT-RNR2  \n",
       "4                           MT    MT-TL1  \n",
       "...                        ...       ...  \n",
       "76116                        5     MAP1B  \n",
       "76117                        5     MAP1B  \n",
       "76118                        5     MAP1B  \n",
       "76119                        5     MAP1B  \n",
       "76120                      NaN       NaN  \n",
       "\n",
       "[76121 rows x 6 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "GRCh_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "27567d46-fdad-46f8-aabf-48c3ecf4230d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Stable ID</th>\n",
       "      <th>Gene Name</th>\n",
       "      <th>Transcription Factor</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>None</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000000005</td>\n",
       "      <td>None</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000000419</td>\n",
       "      <td>None</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000000457</td>\n",
       "      <td>None</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000000460</td>\n",
       "      <td>None</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41318</th>\n",
       "      <td>ENSG00000181404</td>\n",
       "      <td>None</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41319</th>\n",
       "      <td>ENSG00000276581</td>\n",
       "      <td>None</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41320</th>\n",
       "      <td>ENSG00000278848</td>\n",
       "      <td>None</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41321</th>\n",
       "      <td>ENSG00000279782</td>\n",
       "      <td>None</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41322</th>\n",
       "      <td>ENSG00000280267</td>\n",
       "      <td>None</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>41323 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Stable ID Gene Name  Transcription Factor\n",
       "0      ENSG00000000003      None                 False\n",
       "1      ENSG00000000005      None                 False\n",
       "2      ENSG00000000419      None                 False\n",
       "3      ENSG00000000457      None                 False\n",
       "4      ENSG00000000460      None                 False\n",
       "...                ...       ...                   ...\n",
       "41318  ENSG00000181404      None                 False\n",
       "41319  ENSG00000276581      None                 False\n",
       "41320  ENSG00000278848      None                 False\n",
       "41321  ENSG00000279782      None                 False\n",
       "41322  ENSG00000280267      None                 False\n",
       "\n",
       "[41323 rows x 3 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "indexTable"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "42252eaf-3e93-41dc-a002-ba7bb6def776",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene 1</th>\n",
       "      <th>Gene 2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000000005</td>\n",
       "      <td>ENSG00000061656</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000000005</td>\n",
       "      <td>ENSG00000099968</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000000005</td>\n",
       "      <td>ENSG00000104765</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000000005</td>\n",
       "      <td>ENSG00000105383</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000000005</td>\n",
       "      <td>ENSG00000114455</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64001</th>\n",
       "      <td>ENSG00000275774</td>\n",
       "      <td>ENSG00000275774</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64002</th>\n",
       "      <td>ENSG00000276070</td>\n",
       "      <td>ENSG00000278619</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64003</th>\n",
       "      <td>ENSG00000276076</td>\n",
       "      <td>ENSG00000276076</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64004</th>\n",
       "      <td>ENSG00000276644</td>\n",
       "      <td>ENSG00000276644</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64005</th>\n",
       "      <td>ENSG00000280987</td>\n",
       "      <td>ENSG00000280987</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>64006 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                Gene 1           Gene 2\n",
       "0      ENSG00000000005  ENSG00000061656\n",
       "1      ENSG00000000005  ENSG00000099968\n",
       "2      ENSG00000000005  ENSG00000104765\n",
       "3      ENSG00000000005  ENSG00000105383\n",
       "4      ENSG00000000005  ENSG00000114455\n",
       "...                ...              ...\n",
       "64001  ENSG00000275774  ENSG00000275774\n",
       "64002  ENSG00000276070  ENSG00000278619\n",
       "64003  ENSG00000276076  ENSG00000276076\n",
       "64004  ENSG00000276644  ENSG00000276644\n",
       "64005  ENSG00000280987  ENSG00000280987\n",
       "\n",
       "[64006 rows x 2 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Alist = build.list_HuRI()\n",
    "Alist"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "3168bf89-8e3e-49bd-a2c3-eb991574ef3c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Accessing STRING data\n",
      "    Reading data from file...\n",
      "    File read complete\n",
      "    98.07% of the data was removed because it is below the confidence threshold.\n",
      "    2.12% of the data was removed because it wasn't mapped to a stable gene ID.\n",
      "STRING Data Access Complete: 1.89% of the data was used.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene 1</th>\n",
       "      <th>Gene 2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>ENSG00000143761</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>ENSG00000120370</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>ENSG00000157869</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>ENSG00000134287</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>ENSG00000107862</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>225635</th>\n",
       "      <td>ENSG00000100129</td>\n",
       "      <td>ENSG00000149100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>225636</th>\n",
       "      <td>ENSG00000100129</td>\n",
       "      <td>ENSG00000178982</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>225637</th>\n",
       "      <td>ENSG00000100129</td>\n",
       "      <td>ENSG00000114867</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>225638</th>\n",
       "      <td>ENSG00000100129</td>\n",
       "      <td>ENSG00000104408</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>225639</th>\n",
       "      <td>ENSG00000100129</td>\n",
       "      <td>ENSG00000140988</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>225640 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 Gene 1           Gene 2\n",
       "0       ENSG00000004059  ENSG00000143761\n",
       "1       ENSG00000004059  ENSG00000120370\n",
       "2       ENSG00000004059  ENSG00000157869\n",
       "3       ENSG00000004059  ENSG00000134287\n",
       "4       ENSG00000004059  ENSG00000107862\n",
       "...                 ...              ...\n",
       "225635  ENSG00000100129  ENSG00000149100\n",
       "225636  ENSG00000100129  ENSG00000178982\n",
       "225637  ENSG00000100129  ENSG00000114867\n",
       "225638  ENSG00000100129  ENSG00000104408\n",
       "225639  ENSG00000100129  ENSG00000140988\n",
       "\n",
       "[225640 rows x 2 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Alist, _ = build.list_STRING(thresh=900)\n",
    "Alist"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "1dd9a398-1a5a-4bee-b931-415ef6992c47",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Accessing STRING data\n",
      "    Reading data from file...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "KeyboardInterrupt\n",
      "\n"
     ]
    }
   ],
   "source": [
    "Alist = build.list_STRING(thresh=900)\n",
    "Alist"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "80c1abf5-c10f-4249-8dcd-b326a8cec525",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Accessing STRING data\n",
      "    Reading data from file...\n",
      "    File read complete\n"
     ]
    }
   ],
   "source": [
    "thresh=900\n",
    "STRING_PPI=None\n",
    "print(\"Accessing STRING data\")\n",
    "\n",
    "# Set data path\n",
    "data_path = \"/home/jpic/HardwiredGenome-Python/Data/\"\n",
    "\n",
    "# Read and sort raw PPIs\n",
    "if STRING_PPI is None or not isinstance(STRING_PPI, pd.DataFrame):\n",
    "    print(\"    Reading data from file...\")\n",
    "    STRING_downloaded_PPI = f\"{data_path}/Downloaded/STRING/9606.protein.links.v11.5.txt\"\n",
    "    STRING_PPI = pd.read_csv(STRING_downloaded_PPI, delim_whitespace=True)\n",
    "    print(\"    File read complete\")\n",
    "else:\n",
    "    print(\"    STRING data present from previous file read\")\n",
    "\n",
    "# Apply threshold\n",
    "confident_indices = STRING_PPI['combined_score'].astype(int) > thresh\n",
    "P1_confident = STRING_PPI['protein1'][confident_indices].astype(str)\n",
    "P2_confident = STRING_PPI['protein2'][confident_indices].astype(str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "4349143b-7619-40b5-bd6e-48c46fafa49d",
   "metadata": {},
   "outputs": [],
   "source": [
    "P1_confident = P1_confident.str[5:]\n",
    "P2_confident = P2_confident.str[5:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "c080da8f-f016-454c-a7b7-4ea9d990ec35",
   "metadata": {},
   "outputs": [],
   "source": [
    "P1_gene_id, remove1 = build.stable_ID_map(P1_confident.tolist(), True, uni=False)\n",
    "P2_gene_id, remove2 = build.stable_ID_map(P2_confident.tolist(), True, uni=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "c4c07248-5387-4413-b054-0e10df67d476",
   "metadata": {},
   "outputs": [],
   "source": [
    "remove_rows = np.unique(np.concatenate([remove1, remove2]))\n",
    "idxs = np.ones(len(P1_gene_id), dtype=bool)\n",
    "idxs[remove_rows] = False\n",
    "P1_gene_id = np.array(P1_gene_id)[idxs]\n",
    "P2_gene_id = np.array(P2_gene_id)[idxs]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "72dcbb8d-97e3-4076-856f-e5de333ab980",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "81863476-364d-4781-a177-af059ad9b01a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    2.12% of the data was removed because it wasn't mapped to a stable gene ID.\n",
      "STRING Data Access Complete: 1.89% of the data was used.\n"
     ]
    }
   ],
   "source": [
    "# Calculate further removal statistics\n",
    "rem = 1 - len(P1_gene_id) / len(P1_confident)\n",
    "print(f\"    {rem * 100:.2f}% of the data was removed because it wasn't mapped to a stable gene ID.\")\n",
    "\n",
    "# Organize data as a DataFrame\n",
    "A_list = pd.DataFrame({\n",
    "    'Gene 1': P1_gene_id,\n",
    "    'Gene 2': P2_gene_id\n",
    "})\n",
    "\n",
    "# Calculate total usage\n",
    "rem = len(P1_gene_id) / len(STRING_PPI)\n",
    "print(f\"STRING Data Access Complete: {rem * 100:.2f}% of the data was used.\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "7fafb30a-6ac8-46ac-933c-f7fb2f20adc7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Accessing STRING data\n",
      "    Reading data from file...\n",
      "    File read complete\n",
      "    98.07% of the data was removed because it is below the confidence threshold.\n"
     ]
    },
    {
     "ename": "TypeError",
     "evalue": "stable_ID_map() missing 1 required positional argument: 'uni'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[11], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m A \u001b[38;5;241m=\u001b[39m \u001b[43mbuild\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mlist_STRING\u001b[49m\u001b[43m(\u001b[49m\u001b[43mthresh\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m900\u001b[39;49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/HardwiredGenome-Python/HWG/build.py:323\u001b[0m, in \u001b[0;36mlist_STRING\u001b[0;34m(thresh, STRING_PPI)\u001b[0m\n\u001b[1;32m    320\u001b[0m P2_confident \u001b[38;5;241m=\u001b[39m P2_confident\u001b[38;5;241m.\u001b[39mstr[\u001b[38;5;241m5\u001b[39m:]\n\u001b[1;32m    322\u001b[0m \u001b[38;5;66;03m# Map protein IDs to gene IDs\u001b[39;00m\n\u001b[0;32m--> 323\u001b[0m P1_gene_id, remove1 \u001b[38;5;241m=\u001b[39m \u001b[43mstable_ID_map\u001b[49m\u001b[43m(\u001b[49m\u001b[43mP1_confident\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mtolist\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m)\u001b[49m\n\u001b[1;32m    324\u001b[0m P2_gene_id, remove2 \u001b[38;5;241m=\u001b[39m stable_ID_map(P2_confident\u001b[38;5;241m.\u001b[39mtolist(), \u001b[38;5;28;01mTrue\u001b[39;00m)\n\u001b[1;32m    326\u001b[0m \u001b[38;5;66;03m# Remove rows that were unmapped\u001b[39;00m\n",
      "\u001b[0;31mTypeError\u001b[0m: stable_ID_map() missing 1 required positional argument: 'uni'"
     ]
    }
   ],
   "source": [
    "A = build.list_STRING(thresh=900)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "761235a3-8a73-45df-b40e-68cc1d82176b",
   "metadata": {},
   "outputs": [],
   "source": [
    "A"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "64207954-0c0d-40d8-8cc4-98233b63914a",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_path = build.load_data_path()\n",
    "\n",
    "# Read table\n",
    "GRCh38_data_path = os.path.join(data_path, \"Downloaded/Ensemble/GRCh38.p13.tsv\")\n",
    "GRCh = pd.read_csv(GRCh38_data_path, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "7ea900ba-b9a1-4b32-88c5-920705a185f0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>9606</td>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>ENSP00000362111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>9606</td>\n",
       "      <td>ENSG00000000005</td>\n",
       "      <td>ENSP00000362122</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>9606</td>\n",
       "      <td>ENSG00000000419</td>\n",
       "      <td>ENSP00000360644</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>9606</td>\n",
       "      <td>ENSG00000000457</td>\n",
       "      <td>ENSP00000356744</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>9606</td>\n",
       "      <td>ENSG00000000460</td>\n",
       "      <td>ENSP00000286031</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23443</th>\n",
       "      <td>9606</td>\n",
       "      <td>ENSG00000269882</td>\n",
       "      <td>ENSP00000469899</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23444</th>\n",
       "      <td>9606</td>\n",
       "      <td>ENSG00000269883</td>\n",
       "      <td>ENSP00000470831</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23445</th>\n",
       "      <td>9606</td>\n",
       "      <td>ENSG00000269884</td>\n",
       "      <td>ENSP00000470150</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23446</th>\n",
       "      <td>9606</td>\n",
       "      <td>ENSG00000269955</td>\n",
       "      <td>ENSP00000347005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23447</th>\n",
       "      <td>9606</td>\n",
       "      <td>ENSG00000271303</td>\n",
       "      <td>ENSP00000371388</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>23448 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          0                1                2\n",
       "0      9606  ENSG00000000003  ENSP00000362111\n",
       "1      9606  ENSG00000000005  ENSP00000362122\n",
       "2      9606  ENSG00000000419  ENSP00000360644\n",
       "3      9606  ENSG00000000457  ENSP00000356744\n",
       "4      9606  ENSG00000000460  ENSP00000286031\n",
       "...     ...              ...              ...\n",
       "23443  9606  ENSG00000269882  ENSP00000469899\n",
       "23444  9606  ENSG00000269883  ENSP00000470831\n",
       "23445  9606  ENSG00000269884  ENSP00000470150\n",
       "23446  9606  ENSG00000269955  ENSP00000347005\n",
       "23447  9606  ENSG00000271303  ENSP00000371388\n",
       "\n",
       "[23448 rows x 3 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "STRING_ID_map_path = os.path.join(data_path, \"Downloaded/STRING/string_9606_ENSG_ENSP_10_all_T.tsv\")\n",
    "STRING_IDs = pd.read_csv(STRING_ID_map_path, sep='\\t', header=None)\n",
    "STRING_IDs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "e9ef07fa-966c-4713-9e9a-d5e6a8bd3fe9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0        ENSG00000210049\n",
       "1        ENSG00000211459\n",
       "2        ENSG00000210077\n",
       "3        ENSG00000210082\n",
       "4        ENSG00000209082\n",
       "              ...       \n",
       "76116    ENSG00000131711\n",
       "76117    ENSG00000131711\n",
       "76118    ENSG00000131711\n",
       "76119    ENSG00000131711\n",
       "76120    ENSG00000200637\n",
       "Name: Gene stable ID, Length: 76121, dtype: object"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "GRCh['Gene stable ID']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "5376a589-1850-4d5b-9f9e-bdd146c238fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "P1_confident = P1_confident.str[5:]\n",
    "P2_confident = P2_confident.str[5:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "f4683c7b-7d50-4aa7-b5a4-e4bf0afa817d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "187         ENSP00000440005\n",
       "401         ENSP00000356737\n",
       "565         ENSP00000328551\n",
       "945         ENSP00000256682\n",
       "1025        ENSP00000359000\n",
       "                 ...       \n",
       "11938186    ENSP00000436049\n",
       "11938211    ENSP00000248342\n",
       "11938226    ENSP00000416255\n",
       "11938234    ENSP00000220849\n",
       "11938267    ENSP00000341885\n",
       "Name: protein2, Length: 230524, dtype: object"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "P2_confident"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "ae44d0a4-448f-433c-b6d4-4baa23198bd6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "187         ENSP00000000233\n",
       "401         ENSP00000000233\n",
       "565         ENSP00000000233\n",
       "945         ENSP00000000233\n",
       "1025        ENSP00000000233\n",
       "                 ...       \n",
       "11938186    ENSP00000485663\n",
       "11938211    ENSP00000485663\n",
       "11938226    ENSP00000485663\n",
       "11938234    ENSP00000485663\n",
       "11938267    ENSP00000485663\n",
       "Name: protein1, Length: 230524, dtype: object"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "P1_confident"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "fa58e864-10ea-48ba-a788-49a55959d467",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "471caa77-c23a-4fb3-8d4c-8269a5313078",
   "metadata": {},
   "outputs": [],
   "source": [
    "HuRI_PPI1 = pd.read_csv(\"/home/jpic/HardwiredGenome-Python/Data/Downloaded/HuRI/HI-union.tsv\", sep='\\t', header=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "02cabbeb-550b-4ba2-acb3-36953d39cc18",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['ENSG00000000005', 'ENSG00000000005', 'ENSG00000000005', ...,\n",
       "       'ENSG00000276076', 'ENSG00000276644', 'ENSG00000280987'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "HuRI_PPI1[0].values"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
